File: scripts/lib/states.ts
import * as fs from 'fs/promises';
import * as path from 'path';

/**
 * A single state/UT the MLA import pipeline can process. `pdf` is the
 * source list PDF a state coordinator drops in locally; `dataFile` and
 * `imageDir` are derived by convention so config/states.json only has to
 * name the state, its code, and where its PDF lives - nothing else.
 */
export interface StateConfig {
  name: string;
  code: string;
  pdf: string;
  dataFile: string;
  imageDir: string;
}

interface RawStateConfig {
  name: string;
  code: string;
  pdf: string;
}

const DEFAULT_CONFIG_PATH = 'config/states.json';

function withDerivedPaths(raw: RawStateConfig): StateConfig {
  return {
    ...raw,
    dataFile: `data/import/${raw.code}_mlas.json`,
    imageDir: `storage/leaders/${raw.code}`
  };
}

export async function loadStates(configPath: string = DEFAULT_CONFIG_PATH): Promise<StateConfig[]> {
  const absolutePath = path.resolve(process.cwd(), configPath);
  const contents = await fs.readFile(absolutePath, 'utf-8');
  const raw = JSON.parse(contents) as RawStateConfig[];

  if (!Array.isArray(raw)) {
    throw new Error(`${configPath} must contain a JSON array of state entries.`);
  }

  const seenCodes = new Set<string>();
  for (const entry of raw) {
    if (!entry.name || !entry.code || !entry.pdf) {
      throw new Error(`Invalid entry in ${configPath}: each state needs a name, code, and pdf path.`);
    }
    if (seenCodes.has(entry.code)) {
      throw new Error(`Duplicate state code '${entry.code}' in ${configPath}.`);
    }
    seenCodes.add(entry.code);
  }

  return raw.map(withDerivedPaths);
}

export async function getState(code: string, configPath?: string): Promise<StateConfig> {
  const states = await loadStates(configPath);
  const match = states.find((s) => s.code === code);
  if (!match) {
    const known = states.map((s) => s.code).join(', ');
    throw new Error(`Unknown state code '${code}'. Configured codes: ${known}`);
  }
  return match;
}

File: scripts/lib/states.py
"""Shared config loader for the MLA extraction pipeline.

Every script (parse_pdf.py, extract_images.py, match_images.py,
rename_images.py, extract_state_mlas.py) reads state metadata through
this module instead of hardcoding a state name, PDF path, or output
folder. Adding a new state means adding one entry to config/states.json -
never adding a branch to any script.
"""
import argparse
import json
import os

DEFAULT_CONFIG_PATH = "config/states.json"


class UnknownStateError(Exception):
    pass


def load_states(config_path=DEFAULT_CONFIG_PATH):
    with open(config_path, "r", encoding="utf-8") as f:
        raw = json.load(f)

    if not isinstance(raw, list):
        raise ValueError(f"{config_path} must contain a JSON array of state entries.")

    seen_codes = set()
    states = []
    for entry in raw:
        if not entry.get("name") or not entry.get("code") or not entry.get("pdf"):
            raise ValueError(f"Invalid entry in {config_path}: each state needs a name, code, and pdf path.")
        if entry["code"] in seen_codes:
            raise ValueError(f"Duplicate state code '{entry['code']}' in {config_path}.")
        seen_codes.add(entry["code"])

        state = dict(entry)
        state["data_file"] = os.path.join("data", "import", f"{entry['code']}_mlas.json")
        state["image_dir"] = os.path.join("storage", "leaders", entry["code"])
        states.append(state)

    return states


def get_state(code, config_path=DEFAULT_CONFIG_PATH):
    states = load_states(config_path)
    for state in states:
        if state["code"] == code:
            return state
    known = ", ".join(s["code"] for s in states)
    raise UnknownStateError(f"Unknown state code '{code}'. Configured codes: {known}")


def state_arg_parser(description):
    """Standard --state=<code> / --config=<path> CLI for every pipeline script."""
    parser = argparse.ArgumentParser(description=description)
    parser.add_argument("--state", required=True, help="State code from config/states.json (e.g. up, br, ka)")
    parser.add_argument("--config", default=DEFAULT_CONFIG_PATH, help="Path to states config (default: config/states.json)")
    return parser
